HIP/hip-584.md

```diff
@@ -0,0 +1,127 @@
+---
+hip: 584
+title: EVM Module
+author:  Ivan Kavaldzhiev <ivan.kavaldzhiev@limechain.tech>, Mustafa Uzun <mustafa.uzun@limechain.tech>
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Draft
+created: 2022-09-27
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/586
+---
+
+## Abstract
+
+Encapsulating Hedera smart contract EVM specific logic in a library.
+It would expose interfaces that will allow different implementations.
+Mirror node would build ethereum-related REST API on top of the library. 
+Mirror nodes could use the library to simulate a transaction and the output would be used to 
+populate the return value for the REST API that is invoked. 
+Consensus nodes would also use the library and keep their existing logic by providing specific interface implementations.
+
+## Motivation
+
+Extracting common Hedera smart contract EVM logic in a single place
+
+Enabling free read-only queries.
+
+Enabling free gas estimate for a given transaction.
+
+Enabling free transaction simulations.
+
+## Rationale
+
+In order to be possible to both execute and simulate transactions, run by the EVM used inside `hedera-node`,
+we need to extract a module encapsulating all of the EVM logic.
+
+The resulting library would allow us to fulfil the following use cases:
+
+- Execute transactions in a regular flow with committing changes to state
+- Execute transaction simulations that perform read-only operations
+- Execute transaction simulations that perform speculative writes without committing changes to state
+- Execute transaction simulations that perform speculative writes using historical data for transaction replay, 
+without committing changes to state
+
+## User stories
+
+- As a developer, I want to have a library with common Hedera smart contract EVM logic that I can reuse in different projects
+- As a user, I want to perform `eth_call` calls from the Ethereum JSON-RPC standard
+- As a user, I want to perform `eth_gasEstimate` calls from the Ethereum JSON-RPC standard
+
+## Specification
+
+### EVM Library Definition
+
+We want the EVM library to be used by multiple clients like the `hedera-services` and `hedera-mirror-node` repos
+and to be able to be extended for a future use from additional clients.
+
+To facilitate differences, we need to define interfaces, which would be inside the library 
```

Comment:
Should we not be listing the suggested major interfaces and their responsibilities here?

Reply:
We may define the basic interface for interacting with the library. It will be inside `hedera-evm-api` module and `mirror-node` would use it to spin up a simulation of a transaction. The output would be `TransactionProcessingResult`. `hedera-evm-api` on its own side would consume and use the `hedera-evm` implementation module to perform the transaction. The dependency chain would look like this:
`besu` -> `hedera-evm` -> `hedera-evm-api `-> `hedera-node`/`hedera-mirror-web3`

Going further into more details about the other interfaces, IMO, might introduce more specific and unnecessary information for the context of a HIP.

Reply:
Maybe this can be in a design doc committed into the services repo inside the `hedera-evm` repo